### PR TITLE
fix(results): isolate nested member integrity hashes

### PR DESCRIPTION
**Pull request by mldangelo-oai**
## Summary

- Preserve parent artifact hashes when nested archive or wrapper scans emit their own integrity metadata.
- Export child hashes as structured `member_file_hashes` records with raw path segments and deterministic occurrence identities, including explicit partial-prefix records.
- Bound retained member integrity state to 4,096 records and a 2 MiB cumulative encoded-path budget, with total/truncated/omitted summaries.
- Carry the contract through local and streaming JSON, cache round trips, SARIF, and CycloneDX SBOM while redacting credentials from exported member paths.
- Apply the same parent/child attribution to ZIP, TAR, 7z, compressed wrappers, PyTorch ZIP, Keras, OCI layers, ExecuTorch, TorchServe MAR, and compressed Joblib payloads.

## Why

Parent artifacts and nested members previously shared the same `file_hashes` metadata key. Generic merges and archive-specific restore paths could therefore report a child digest as the parent artifact digest. Earlier member identities were also delimiter-sensitive, private tracking was not fully bounded, and some exporter paths could retain credential-bearing member provenance.

## Security implications

This change fixes integrity attribution and export redaction without weakening detections or fail-closed behavior. Malicious nested findings still propagate. Complete and partial member hashes remain child-scoped, skipped members do not receive synthetic hashes, and nested diagnostic snapshots no longer duplicate the bounded member-hash payload.

## Review fixes included

- Removed the redundant OCI shallow restore that could overwrite the preserved parent digest.
- Preserved current-main Keras metadata precedence while carrying nested hash summaries.
- Kept PyTorch nested merges linear and excluded already-aggregated hash maps from diagnostic snapshots.
- Hash compressed Joblib wrappers as parents and their original decompressed payloads as members, even when analysis uses a sanitized control buffer.
- Redact member identities and path fields in both typed and legacy SBOM generation.
- Added collision, duplicate, partial-prefix, long-path budget, cache, streaming, exporter, and archive-specific regressions.

## Validation on `2135668668ea02f42b6461d977ac2fab38c7f974`

```bash
uv run --frozen ruff format --check modelaudit/ packages/modelaudit-picklescan/src packages/modelaudit-picklescan/tests tests/
# 422 files already formatted

uv run --frozen ruff check modelaudit/ packages/modelaudit-picklescan/src packages/modelaudit-picklescan/tests tests/
# All checks passed

MYPY_CACHE_DIR=/tmp/modelaudit-pr1666-mypy uv run --frozen mypy modelaudit/ packages/modelaudit-picklescan/src packages/modelaudit-picklescan/tests tests/
# Success: no issues found in 477 source files

PROMPTFOO_DISABLE_TELEMETRY=1 uv run --frozen pytest -n auto -m "not slow and not integration" --maxfail=1
# 20,805 passed, 785 skipped, 41 warnings
```

After additively merging PR #1679, its exact picklescan regression passed 24/24 before the full gates above.

Additional focused validation covered parent/member hashes, duplicate and delimiter-collision paths, bounded record/path state, partial prefixes, OCI and Keras restore behavior, PyTorch nested summaries and performance, compressed Joblib payload identity, cache reads, streaming, SARIF, and both SBOM implementations. Independent owner, security, contract, merge-conflict, and validation lanes reported no remaining P0-P3 findings.

## Residual risk

Detailed member integrity records are retained up to 4,096 records and a 2 MiB cumulative encoded-path budget. Additional records are still counted and reported as omitted/truncated, but their individual hashes are intentionally unavailable to downstream exports.

### CODE REVIEW

**Comment by mldangelo-oai** [comment]
@codex review

**Comment by mldangelo-oai** [comment]
@codex address these exact-head blockers before merge:

- **P1:** `merge_member_result()` deep-copies the accumulated member-hash map for every member, producing O(n²) work. Exact-head probes took 0.69s/2.79s/18.3s for 1k/2k/5k unique members and 24.5s for 5k duplicate names, versus 0.005s for direct O(1) insertion.
- **P2:** member identity keys are lossy. Same-name/same-content duplicates overwrite; `strip()` collapses legal whitespace-distinct names; nested `nested.zip -> b.pkl` collides with literal `nested.zip:b.pkl`; synthetic `#2` also collides with legal path text. Preserve raw path segments plus deterministic occurrence identity independent of hash equality.
- **P3:** the cache round-trip test never proves the second scan was a cache hit.

Bound retained/exported integrity records, avoid per-member full copies, use reversible collision-free identity, and add ZIP/nested/cache plus other archive-format controls. Independent report: `/Users/mdangelo/modelaudit-pr-swarm/outputs/pr-1666-independent-review-20260611.md`.

**Comment by mldangelo-oai** [comment]
Addressed the latest blocker feedback in `292ee38dffa9e0bd56146b18bfb7dfb137410f61`:

- P1: member hash insertion no longer deep-copies the accumulated map per member. It uses private occurrence counters with O(1) insertion and caps retained exports at 4096 records, with total/truncated/omitted summary fields for bounded adversarial archives.
- P2: member identity is now structured and reversible: raw `path_segments` plus `occurrence` are stored in each record and encoded into the map key. Duplicate same-name/same-content members, whitespace-distinct names, literal `:` paths, nested archive segments, and legal `#2` names no longer collide.
- P3: the cache regression now asserts the second scan increments `cache_hits`, so the test proves member integrity metadata survives an actual cache read.

Additional controls added for PyTorch ZIP JSON/SARIF/SBOM, generic ZIP duplicate/nested/literal-path cases with a malicious pickle member, tar child scoping, malformed/skipped/partial/inconclusive members, streaming scans, and the pinned real-HF model.

Validation on the new head:

```bash
PROMPTFOO_DISABLE_TELEMETRY=1 MODELAUDIT_RUN_REAL_HF=1 uv run pytest tests/test_regular_scan_hash.py::TestRegularScanContentHash::test_real_hf_bert_large_rust_model_member_hash_is_namespaced -q
# 1 passed in 183.60s

PROMPTFOO_DISABLE_TELEMETRY=1 uv run pytest tests/test_models.py tests/test_regular_scan_hash.py tests/test_streaming_scan.py tests/scanners/test_pytorch_zip_scanner.py -q
# 1019 passed, 1 skipped, 6 warnings in 420.67s

PROMPTFOO_DISABLE_TELEMETRY=1 uv run pytest -n auto -m "not slow and not integration" --maxfail=1 -q
# 17375 passed, 1292 skipped, 40 warnings in 775.09s

uv run ruff check modelaudit/ packages/modelaudit-picklescan/src packages/modelaudit-picklescan/tests tests/
uv run ruff format --check modelaudit/ packages/modelaudit-picklescan/src packages/modelaudit-picklescan/tests tests/
uv run mypy modelaudit/ packages/modelaudit-picklescan/src packages/modelaudit-picklescan/tests tests/
git diff --check
# clean
```

@codex review

**Comment by mldangelo-oai** [comment]
Addressed the Codex P2 review threads in `a719fab7a7227992deac024adae8db2a8365cbda`:

- Rebuild/counter safety: `merge()` now skips derived member-hash private occurrence and retained-count metadata from child results, so nested child merges cannot reset parent counters or overwrite an existing occurrence.
- SARIF truncation summary: SARIF artifact properties now include `memberFileHashesTotal`, `memberFileHashesTruncated`, and `memberFileHashesOmitted` alongside retained `memberFileHashes`.
- PyTorch nested ZIP metadata reset: `_merge_nested_zip_result()` now preserves the member-hash total/truncated/omitted summary when restoring parent metadata after nested scan merging.

Additional regressions cover the counter-clobber overwrite case, SARIF/SBOM bounded-summary export, and PyTorch nested ZIP summary preservation. Latest local validation is reflected in the PR body, including the pinned HF reproduction on `a719fab7`.

@codex review

**Comment by mldangelo-oai** [comment]
Independent exact-head gate for `a719fab7a7227992deac024adae8db2a8365cbda`: **REQUEST CHANGES**.

**P2, validated at 96% confidence:** retained public member-hash records are capped, but private identity state is not. `_add_member_file_hash_record()` inserts every unique normalized path into `member_file_hash_occurrences` before checking the 4,096-record cap. A 10,000-member adversarial probe correctly retained 4,096 records and reported 5,904 omitted, but still retained all 10,000 private occurrence entries and peaked at 4.35 MB.

Move the cap check before path normalization/occurrence tracking while preserving scalar total/omitted accounting, and add a unique-path truncation regression that proves both public and private structures stay bounded.

Other adversarial coverage held: nested TAR to PyTorch ZIP isolation, duplicate/same-content records, path aliases, traversal rejection, zero-byte inconclusive behavior, filtered owners, failed-child accounting, cache/SARIF/SBOM round trips, and streaming paths. Focused tests passed 12/12, `tests/test_models.py` passed 77/77, Ruff/format/mypy passed, exact-head CI is green, and all prior threads are resolved.

**Comment by mldangelo-oai** [comment]
@codex review

**Comment by mldangelo-oai** [comment]
Addressed the independent exact-head gate in `499738158231dc1e879f4b209115e35968397189`:\n\n- Private member occurrence state is now bounded with the public retained-record cap: once `MAX_MEMBER_FILE_HASH_RECORDS` is reached, `_add_member_file_hash_record()` increments total/omitted counters and returns before path normalization or private occurrence-key insertion.\n- Added a unique-name truncation regression proving retained public records and private `member_file_hash_occurrences` both stay at the cap while total/omitted accounting remains exact.\n\nLatest local validation on this exact head:\n\n```bash\nPROMPTFOO_DISABLE_TELEMETRY=1 uv run pytest tests/test_models.py::test_merge_member_result_bounds_private_occurrence_state_for_unique_truncated_members tests/test_models.py::test_merge_member_result_bounds_retained_member_hashes_without_quadratic_copying -q\n# 2 passed\n\nPROMPTFOO_DISABLE_TELEMETRY=1 uv run pytest tests/test_models.py tests/test_regular_scan_hash.py tests/test_streaming_scan.py tests/scanners/test_pytorch_zip_scanner.py tests/scanners/test_compressed_scanner.py -q\n# 1099 passed, 1 skipped, 6 warnings\n\nuv run ruff format modelaudit/ packages/modelaudit-picklescan/src packages/modelaudit-picklescan/tests tests/\n# 419 files left unchanged\n\nuv run ruff check --fix modelaudit/ packages/modelaudit-picklescan/src packages/modelaudit-picklescan/tests tests/\n# All checks passed\n\nuv run mypy modelaudit/ packages/modelaudit-picklescan/src packages/modelaudit-picklescan/tests tests/\n# Success: no issues found in 474 source files\n\nPROMPTFOO_DISABLE_TELEMETRY=1 uv run pytest -n auto -m "not slow and not integration" --maxfail=1\n# 17381 passed, 1292 skipped, 40 warnings\n```\n\n@codex review

**Comment by mldangelo-oai** [comment]
Synchronized current `main` and addressed the validated exact-head blockers in `de00058654f7101bcdfa9b4fc3412f66a82d9618`:

- removed the OCI parent-hash overwrite and preserved current-main Keras metadata behavior;
- added structured multi-segment identities, partial-prefix records, a 4,096-record/2 MiB path budget, and bounded PyTorch nested diagnostics;
- corrected compressed Joblib parent/decompressed-payload attribution;
- redacted member provenance in both SBOM paths;
- kept PyTorch nested merging linear and added direct regressions for each issue.

Final local gates: Ruff format/check clean, full mypy clean across 477 files, and the non-slow/non-integration suite passed with 20,805 tests and 785 documented skips. Internal owner, security, contract, merge-conflict, and validation lanes report no remaining P0-P3 findings.

@codex review

**Comment by mldangelo-oai** [comment]
Fleet sequencing note: please continue exact-head review, validation, metadata cleanup, and CI on `de000586`, but hold the final merge until PR #1679 has landed. #1679 is the Nightly repair and is currently in its last four long Python lanes. Once #1679 reports MERGED, fetch the new `origin/main`, merge it additively into this branch (no rebase or force-push), re-run the exact-new-head gates, then proceed with the normal squash/admin-review-only merge path. Keep the goal active while waiting.

**Comment by mldangelo-oai** [comment]
Fleet release: Nightly repair PR #1679 is MERGED as `b83dff38026574e255756613d46f65f0010a11b6` (2026-06-13 09:07:12Z). Fetch `origin/main` now and merge that exact current main additively into this published branch; do not rebase or force-push. Reconcile any overlap according to current behavior, re-run the final review/simplification/thread/metadata gates, push, and require exact-new-head CI before squash merge. If `main` advances again before your merge, repeat the normal merge-from-main and exact-head gate. The prior hold is released; continue the goal through verified merge and post-merge main health.

**Comment by mldangelo-oai** [comment]
Serial landing update: #1673 is the next release gate because it is a one-test fail-closed regression with a fully green pre-sync run. Continue your current-main sync, review, simplification, and CI work, but hold the final merge until #1673 reports MERGED. This avoids making #1673 restart against an unrelated large feature branch. Keep the goal active; once #1673 lands, merge the newest main again if needed and complete your exact-head landing cycle.

**Comment by mldangelo-oai** [comment]
Post-#1679 synchronization is complete on exact head `2135668668ea02f42b6461d977ac2fab38c7f974`.

- merged current `main` (`b83dff38026574e255756613d46f65f0010a11b6`) additively with no conflicts;
- the #1679 picklescan regression passed 24/24;
- Ruff format/check passed;
- mypy passed across 477 source files;
- the full non-slow/non-integration suite passed: 20,805 passed, 785 documented skips, 41 warnings.

All three existing review threads remain resolved and the internal owner/security/contract/simplification/validation lanes report no P0-P3 blocker. Requesting a fresh review and CI decision on this exact head.

@codex review

**Comment by mldangelo-oai** [comment]
Merged as `50e2df398a8165f03cb8435fef30be0e7474a75c`.

Post-merge audit:

- release-please, benchmarks, Docker, documentation, CodeQL Security Analysis, and CodeQL completed successfully;
- Python CI reproduced the same two Python 3.11 false-positive test failures seen on both preceding `main` SHAs (`c94fc923` and `b83dff38`): `test_bert_model_no_false_positive_executables` and `test_comprehensive_gpt2_model_scan`;
- #1666 did not modify those tests or the detector paths implicated by them, while its full local suite and exact-head PR matrix passed.

This is verified pre-existing `main` noise, not a regression introduced by #1666, so no unrelated follow-up change was made.
